### PR TITLE
implement BitMap

### PR DESCRIPTION
**Pull request by djy0**
relate to #16

### CODE REVIEW

**Comment by djy0** [comment]
just add a draft to check whether I am on the correct path

**Comment by djy0** [comment]
@jolestar PTAL

**Comment by jolestar** [comment]
I think we should use the bit to reduce storage usage, like https://github.com/starcoinorg/starcoin/blob/2d55f960f92cad2333bf93b2b5b152403cc5f17f/contrib-contracts/modules/MerkleDistributor.move#L129

Another approach is waiting for the Table implementation.

File: sources/BitMap.move [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+                item.bits = item.bits | mask;
+                return
+            };
+            i = i + 1
+        };
+        Vector::push_back(&mut bitMap.data, Item { index: itemIndex, bits: mask })
+    }
+
+    public fun unset(bitMap: &mut BitMap, index: u128) {
+        let itemIndex = index >> 7;
```
Comment (jolestar) [inline sources/BitMap.move:55]: Why `itemIndex = index >> 7`, the bits is u128.

File: sources/BitMap.move [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+        assert!(get(&mut bitMap, 101) == false, 1);
+        unset(&mut bitMap, 100);
+        assert!(get(&mut bitMap, 100) == false, 1);
+        assert!(get(&mut bitMap, 101) == false, 1);
+        assert!(get(&mut bitMap, 88) == true, 1);
+        unset(&mut bitMap, 88);
+        assert!(get(&mut bitMap, 88) == false, 1);
+
+        // test `2 ** 128 - 1`
+        assert!(get(&mut bitMap, 340282366920938463463374607431768211455) == false, 1);
```
Comment (jolestar) [inline sources/BitMap.move:97]: How many items are in the vector, if the index is so big? Please help to write an integration test to eval the gas consumption.

Comment (djy0) [reply sources/BitMap.move:97]: Sure

File: sources/BitMap.move (hunk 1 above)
Comment (djy0) [reply sources/BitMap.move]: I am using the upper 121 bit of key to discriminate items, and use a u128 `bits` to store the items

File: sources/BitMap.move [hunk 3]
```diff
@@ -3,56 +3,66 @@ module SFC::BitMap{
...
         let i = 0;
         let v = &mut bitMap.data;
         let len = Vector::length(v);
         while (i < len) {
             let item = Vector::borrow_mut(v, i);
-            if (item.index == index) {
-                item.isSet = false;
+            if (item.index == itemIndex) {
+                // we use bit `xor` with `2 ** 128 - 1` to emulate bit invert op here 
+                item.bits = item.bits & (340282366920938463463374607431768211455 ^ mask);
```
[the remark this answers is not in the document]
Comment (djy0) [reply sources/BitMap.move]: done

File: sources/BitMap.move (hunk 2 above)
Comment (djy0) [reply sources/BitMap.move]: there won't be much item in vector even if index is big, since we only push at most one item for one input

Comment (djy0) [reply sources/BitMap.move]: the gas used is 275802

**Review by jolestar — APPROVED** [review]
LGTM

**Comment by jolestar** [comment]
@lerencao Please help to review

**Pull request merged**